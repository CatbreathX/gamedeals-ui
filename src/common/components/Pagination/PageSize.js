import {
  FormControl, InputLabel, MenuItem, Select,
} from '@mui/material';
import PropTypes from 'prop-types';

export const PageSize = ({
  labelId,
  pageSize,
  onPageSizeChange,
}) => (
  <FormControl>
    <InputLabel id={labelId}>Page Size</InputLabel>
    <Select
      role="combobox"
      label="Page Size"
      labelId={labelId}
      onChange={onPageSizeChange}
      sx={{ width: '80px' }}
      value={pageSize}
    >
      <MenuItem value={10}>10</MenuItem>
      <MenuItem value={25}>25</MenuItem>
      <MenuItem value={60}>60</MenuItem>
    </Select>
  </FormControl>
);

PageSize.propTypes = {
  labelId: PropTypes.string.isRequired,
  pageSize: PropTypes.number.isRequired,
  onPageSizeChange: PropTypes.func.isRequired,
};
