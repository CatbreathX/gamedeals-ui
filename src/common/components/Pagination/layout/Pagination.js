import styled from '@emotion/styled';
import { LARGE } from 'common/layout/constants';

export const PaginationContainer = styled.div`
  align-items: center;
  justify-content: flex-start;
  display: inline-flex;
  flex-direction: row;
  column-gap: 24px;

  @media screen and (max-width: ${LARGE}) {
    flex-direction: row;
  }
`;

export const Div = styled.div`
`;
