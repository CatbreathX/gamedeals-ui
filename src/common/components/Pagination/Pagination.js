import { Button } from '@mui/material';
import { Div, PaginationContainer } from 'common/components/Pagination/layout/Pagination';
import PropTypes from 'prop-types';
import { PageSize } from 'common/components/Pagination/PageSize';
import NavigateNextIcon from '@mui/icons-material/NavigateNext';
import NavigateBeforeIcon from '@mui/icons-material/NavigateBefore';

export const Pagination = ({
  numberOfRows,
  onNextPage,
  onPageSizeChange,
  onPreviousPage,
  pageNumber,
  pageSize,
  totalNumberOfPages,
}) => {
  const rowNumberStart = (pageNumber * numberOfRows) - numberOfRows + 1;
  const rowNumberEnd = rowNumberStart + numberOfRows - 1;
  const isNextPageEnabled = totalNumberOfPages > pageNumber;
  const isPreviousPageEnabled = pageNumber !== 1;

  const onPreviousPageClicked = (e) => {
    e.preventDefault();
    onPreviousPage(pageNumber - 1);
  };

  const onNextPageClicked = (e) => {
    e.preventDefault();
    onNextPage(pageNumber + 1);
  };

  const onPageSizeChanged = (e) => {
    e.preventDefault();
    onPageSizeChange(e.target.value);
  };

  return (
    <PaginationContainer role="navigation" aria-label="Pagination for search results">
      <PageSize labelId="top" onPageSizeChange={onPageSizeChanged} pageSize={pageSize} />
      <Div aria-label="Page Number">
        Page:
        <span>
          {' '}
          {pageNumber}
        </span>
      </Div>
      <Div aria-label="Start row number to end row number">
        {`Rows ${rowNumberStart} to ${rowNumberEnd}`}
      </Div>
      <Div>
        <Button
          sx={{ fontSize: 18 }}
          aria-label="Goto previous page"
          onClick={onPreviousPageClicked}
          disabled={!isPreviousPageEnabled}
        >
          <NavigateBeforeIcon />
        </Button>
        <Button
          sx={{ zIndex: 9999 }}
          aria-label="Goto next page"
          onClick={onNextPageClicked}
          disabled={!isNextPageEnabled}
        >
          <NavigateNextIcon />
        </Button>
      </Div>
    </PaginationContainer>
  );
};

Pagination.propTypes = {
  numberOfRows: PropTypes.number.isRequired,
  onNextPage: PropTypes.func.isRequired,
  onPageSizeChange: PropTypes.func.isRequired,
  onPreviousPage: PropTypes.func.isRequired,
  pageNumber: PropTypes.number.isRequired,
  pageSize: PropTypes.number.isRequired,
  totalNumberOfPages: PropTypes.number.isRequired,
};
