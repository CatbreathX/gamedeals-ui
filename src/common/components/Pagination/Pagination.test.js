import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Pagination } from 'common/components/Pagination/Pagination';

describe('Pagination', () => {
  test('should disable previous page when on first page', () => {
    renderComponent({ pageNumber: 1 });
    expect(screen.getByRole('button', { name: 'Goto previous page' })).toBeDisabled();
  });

  test('should render enable next page button when more pages of data', () => {
    renderComponent();
    expect(screen.getByRole('button', { name: 'Goto next page' })).toBeEnabled();
  });

  test('should disable next page button when on last page', () => {
    renderComponent({ totalNumberOfPages: 1 });
    expect(screen.getByRole('button', { name: 'Goto next page' })).toBeDisabled();
  });

  test('should enable previous and next page buttons when previous and next pages exist', () => {
    renderComponent({
      pageNumber: 2,
      totalNumberOfPages: 3,
    });

    expect(screen.getByRole('button', { name: 'Goto previous page' })).toBeEnabled();
    expect(screen.getByRole('button', { name: 'Goto next page' })).toBeEnabled();
  });

  test('should invoke nextPage handler when user clicks on next page button', () => {
    const mockHandler = jest.fn();
    renderComponent({
      totalNumberOfPages: 2,
      onNextPage: mockHandler,
    });
    const nextPageButton = screen.getByRole('button', { name: 'Goto next page' });
    userEvent.click(nextPageButton);
    expect(mockHandler).toHaveBeenCalledWith(2);
  });

  test('should invoke previousPage handler when user clicks on previous page button', () => {
    const mockHandler = jest.fn();
    renderComponent({
      pageNumber: 2,
      totalNumberOfPages: 2,
      onPreviousPage: mockHandler,
    });
    const previousPageButton = screen.getByRole('button', { name: 'Goto previous page' });
    userEvent.click(previousPageButton);
    expect(mockHandler)
      .toHaveBeenCalledWith(1);
  });

  test('should display correct page number', () => {
    renderComponent({
      pageNumber: 1,
      totalNumberOfPages: 1,
    });

    expect(screen.getByLabelText('Page Number')).toHaveTextContent('Page: 1');
  });

  test('should display correct start to end row numbers', () => {
    renderComponent({
      pageNumber: 1,
      totalNumberOfPages: 1,
      numberOfRows: 25,
    });

    expect(screen.getByLabelText('Start row number to end row number'))
      .toHaveTextContent('1 to 25');
  });
});

const noop = () => {
};

const renderComponent = ({
  numberOfRows = 25,
  onNextPage = noop,
  onPageSizeChange = noop,
  onPreviousPage = noop,
  pageNumber = 1,
  pageSize = 25,
  totalNumberOfPages = 99,
} = {}) => {
  render(
    <Pagination
      numberOfRows={numberOfRows}
      onNextPage={onNextPage}
      onPageSizeChange={onPageSizeChange}
      onPreviousPage={onPreviousPage}
      pageNumber={pageNumber}
      pageSize={pageSize}
      totalNumberOfPages={totalNumberOfPages}
    />,
  );
};
