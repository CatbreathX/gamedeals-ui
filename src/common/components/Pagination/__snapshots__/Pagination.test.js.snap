// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination should render correctly 1`] = `
<DocumentFragment>
  <div
    aria-label="Pagination for search results"
    class="css-180cxc5-PaginationContainer eom5tjc1"
    role="navigation"
  >
    <div
      class="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
    >
      <label
        class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        id="top"
      >
        Page Size
      </label>
      <div
        class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-msnvm3-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
        role="combobox"
      >
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-labelledby="top"
          class="MuiSelect-select MuiSelect-outlined MuiOutlinedInput-input MuiInputBase-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
          role="button"
          tabindex="0"
        >
          25
        </div>
        <input
          aria-hidden="true"
          class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
          tabindex="-1"
          value="25"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
          data-testid="ArrowDropDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-1in441m"
          >
            <span>
              Page Size
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      aria-label="Page Number"
      class="css-uru5fq-Div eom5tjc0"
    >
      Page:
      <span>
         1
      </span>
    </div>
    <div
      aria-label="Start row number to end row number"
      class="css-uru5fq-Div eom5tjc0"
    >
      Rows 1 to 25
    </div>
    <div
      class="css-uru5fq-Div eom5tjc0"
    >
      <button
        aria-label="Goto previous page"
        class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root Mui-disabled  css-lzr00z-MuiButtonBase-root-MuiButton-root"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="NavigateBeforeIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.41 7.41 14 6l-6 6 6 6 1.41-1.41L10.83 12z"
          />
        </svg>
      </button>
      <button
        aria-label="Goto next page"
        class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-6618to-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="NavigateNextIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M10 6 8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
